src/pybind/mgr/dashboard/frontend/src/app/ceph/rgw/rgw-multisite-details/rgw-multisite-details.component.ts

```diff
@@ -0,0 +1,224 @@
+        this.nodes = this.abstractTreeData(data);
+      });
+  }
+
+  ngOnDestroy() {
+    this.sub.unsubscribe();
+  }
+
+  getRealmTree(realm: any) {
+    let rootNodes = {};
+    rootNodes['id'] = realm.id;
+    this.realmIds.push(realm.id);
+    rootNodes['name'] = realm.name + ' (realm)';
+    rootNodes['info'] = realm;
+    rootNodes['is_default'] = realm.name === this.defaultsInfo['default_realm'] ? true : false;
+    return rootNodes;
+  }
+
+  getZonegroupTree(zonegroup: RgwZonegroup, realm?: RgwRealm) {
+    let firstChildNodes = {};
+    firstChildNodes['id'] = zonegroup.id;
+    firstChildNodes['name'] = zonegroup.name + ' (zonegroup)';
+    firstChildNodes['info'] = zonegroup;
+    firstChildNodes['is_default'] =
+      zonegroup.name === this.defaultsInfo['default_zonegroup'] ? true : false;
+    firstChildNodes['is_master'] = zonegroup.is_master;
+    firstChildNodes['parent'] = realm ? realm.name : '';
+    return firstChildNodes;
+  }
+
+  getZoneTree(zonegroup: RgwZonegroup, zone: any, realm?: RgwRealm) {
+    let secondChildNodes = {};
+    secondChildNodes['id'] = zone.id;
+    secondChildNodes['name'] = zone.name + ' (zone)';
+    secondChildNodes['info'] = zone;
+    secondChildNodes['parent'] = zonegroup.name;
+    secondChildNodes['second_parent'] = realm ? realm.name : '';
+    secondChildNodes['is_default'] = zone.name === this.defaultsInfo['default_zone'] ? true : false;
+    secondChildNodes['is_master'] = zonegroup.master_zone === zone.id ? true : false;
+    return secondChildNodes;
+  }
+
+  private abstractTreeData(data: any): any[] {
+    let allNodes: any[] = [];
+    let rootNodes = {};
+    let firstChildNodes = {};
+    let allFirstChildNodes = [];
+    let secondChildNodes = {};
+    let allSecondChildNodes: {}[] = [];
+    this.realms = data[0] ? data[0] : [];
+    this.zonegroups = data[1] ? data[1] : [];
+    this.defaultsInfo = data[2] ? data[2] : [];
+    if (this.realms.length > 0) {
+      for (const realm of this.realms) {
+        rootNodes = this.getRealmTree(realm);
+        for (const zonegroup of this.zonegroups) {
+          if (zonegroup.realm_id === realm.id) {
+            firstChildNodes = this.getZonegroupTree(zonegroup, realm);
+            for (const zone of zonegroup.zones) {
+              secondChildNodes = this.getZoneTree(zonegroup, zone, realm);
+              allSecondChildNodes.push(secondChildNodes);
+              secondChildNodes = {};
+            }
+            firstChildNodes['children'] = allSecondChildNodes;
+            allSecondChildNodes = [];
+            allFirstChildNodes.push(firstChildNodes);
+            firstChildNodes = {};
+          }
+        }
+        rootNodes['children'] = allFirstChildNodes;
+        allNodes.push(rootNodes);
+        firstChildNodes = {};
+        secondChildNodes = {};
+        rootNodes = {};
+        allFirstChildNodes = [];
+        allSecondChildNodes = [];
+      }
+    }
+    if (this.zonegroups.length > 0) {
```

Comment:
Rather than building all the tree upfront, why not using the [async API](https://angular2-tree.readme.io/docs/async-data-1)? With a URL scheme like `/rgw/real/<realm_id>/zonegroup/<zg_id>/group/<group_id>` it should be easy to build (it also allows to [customize the field names](https://angular2-tree.readme.io/docs/custom-fields-1)).

Reply:
Ok, the core of my suggestion was to try to avoid any complex data processing in the front-end, so if we're not going to leverage express the RGW hierarchy through an HTTP URL hierarchy, why not doing this in the back-end side and just dump all these tree so it's ready for consumption by the angular2-tree component?